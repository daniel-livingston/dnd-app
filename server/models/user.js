const passportLocalMongoose = require("passport-local-mongoose");
const mongoose = require("mongoose");
const passwordValidator = require("password-validator");
const validator = require("validator");

const userSchema = new mongoose.Schema({
	email: {
		type: String,
		trim: true,
		lowercase: true,
		required: true,
		unique: true,
		validate(value) {
			if (!validator.isEmail(value)) {
				throw new Error("Not a valid email");
			}
		},
	},
	dateCreated: {
		type: Date,
		default: Date.now(),
	},
});

userSchema.methods.toJSON = function () {
	const user = this;
	const userObject = user.toObject();
	delete userObject.attempts;
	delete userObject.last;
	delete userObject.hash;
	delete userObject.salt;
	delete userObject.__v;
	return userObject;
};

const passwordSchema = new passwordValidator();
passwordSchema
	.is()
	.min(8)
	.is()
	.max(50)
	.has()
	.uppercase()
	.has()
	.lowercase()
	.has()
	.digits(1)
	.has()
	.not()
	.spaces();

userSchema.plugin(passportLocalMongoose, {
	passwordValidator(password, cb) {
		if (!passwordSchema.validate(password)) {
			const errors = passwordSchema.validate(password, { list: true });
			return cb(errors);
		}
		cb();
	},
	usernameQueryFields: ["email"],
	limitAttempts: true,
	maxAttempts: 10,
});

const User = mongoose.model("User", userSchema);

module.exports = User;
