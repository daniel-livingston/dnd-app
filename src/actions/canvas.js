export const SAVE_CANVAS = "SAVE_CANVAS";
export const saveCanvas = (canvas) => ({
	type: SAVE_CANVAS,
	canvas: canvas,
});

export const SET_BRUSH_COLOR = "SET_BRUSH_COLOR";
export const setBrushColor = (brushColor) => ({
	type: SET_BRUSH_COLOR,
	brushColor,
});

export const SET_IS_DRAWING = "SET_IS_DRAWING";
export const setIsDrawing = (isDrawing) => ({
	type: SET_IS_DRAWING,
	isDrawing,
});

export const SET_LINE_WIDTH = "SET_LINE_WIDTH";
export const setLineWidth = (lineWidth) => ({
	type: SET_LINE_WIDTH,
	lineWidth,
});
