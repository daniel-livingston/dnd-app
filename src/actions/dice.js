export const ROLL_DICE = "ROLL_DICE";
export const rollDice = (roll) => ({
	type: ROLL_DICE,
	roll,
});

export const UPDATE_DICE_NUMBER = "UPDATE_DICE_NUMBER";
export const updateDiceNumber = (number) => ({
	type: UPDATE_DICE_NUMBER,
	number,
});

export const UPDATE_DICE_SIDES = "UPDATE_DICE_SIDES";
export const updateDiceSides = (sides) => ({
	type: UPDATE_DICE_SIDES,
	sides,
});

export const UPDATE_DICE_MODIFIER = "UPDATE_DICE_MODIFIER";
export const updateDiceModifier = (modifier) => ({
	type: UPDATE_DICE_MODIFIER,
	modifier,
});

export const SET_HISTORY = "SET_HISTORY";
export const setHistory = (history) => ({
	type: SET_HISTORY,
	history,
});

export const ADD_TO_HISTORY = "ADD_TO_HISTORY";
export const addToHistory = (roll) => ({
	type: ADD_TO_HISTORY,
	roll,
});
