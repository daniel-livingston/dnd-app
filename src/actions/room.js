export const JOIN_ROOM = "JOIN_ROOM";
export const joinRoom = (roomData) => ({
	type: JOIN_ROOM,
	data: roomData,
});

export const LEAVE_ROOM = "LEAVE_ROOM";
export const leaveRoom = () => ({
	type: LEAVE_ROOM,
});
