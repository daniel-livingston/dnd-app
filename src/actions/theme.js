const SET_THEME = "SET_THEME";
const setTheme = (theme) => ({
	type: SET_THEME,
	theme,
});

export { SET_THEME, setTheme };
