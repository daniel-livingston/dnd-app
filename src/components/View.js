import React from "react";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import PropTypes from "prop-types";
import useTitle from "../hooks/useTitle";

const propTypes = {
	title: PropTypes.string.isRequired,
	description: PropTypes.string.isRequired,
};

export const View = (props) => {
	useTitle(props.title);

	return (
		<div className='content'>
			<div className='content__header'>
				<div className='content__header__info'>
					<h1 className='content__header__title'>{props.title}</h1>
					<p className='content__header__description'>{props.description}</p>
				</div>
				{props.isLoggedIn ? (
					<Link className='link' to='/'>
						{props.username}
					</Link>
				) : (
					<Link className='link' to='/login'>
						Log in
					</Link>
				)}
			</div>
			<div className='content__body'>{props.children}</div>
		</div>
	);
};

View.propTypes = propTypes;

const mapStateToProps = (state) => ({
	isLoggedIn: state.user.isLoggedIn,
	username: state.user.username,
});

export default connect(mapStateToProps)(View);
