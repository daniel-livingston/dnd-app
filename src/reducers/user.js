import { SET_USER } from "../actions/user";

export const defaultState = {
	isLoggedIn: false,
};

const userReducer = (state = defaultState, action) => {
	switch (action.type) {
		case SET_USER: {
			return {
				...action.user,
				isLoggedIn: true,
			};
		}
		default: {
			return state;
		}
	}
};

export default userReducer;
