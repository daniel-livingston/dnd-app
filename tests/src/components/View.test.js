import React from "react";
import { mount, shallow } from "enzyme";
import { MemoryRouter, Link } from "react-router-dom";
import { Provider } from "react-redux";
import store from "../../../src/store/configureStore";
import View, { View as DisconnectedView } from "../../../src/components/View";
import { setUser } from "../../../src/actions/user";

const wrappedView = (
	<Provider store={store}>
		<MemoryRouter>
			<View title='Test' description='Tests the view component.' />
		</MemoryRouter>
	</Provider>
);

describe("<View>", () => {
	test("renders correctly", () => {
		const wrapper = shallow(
			<DisconnectedView title='Test title' description='Test description' />
		);
		expect(wrapper).toMatchSnapshot();
	});

	test("renders log in link by default", () => {
		const wrapper = mount(wrappedView);
		expect(wrapper.find(Link).prop("to")).toBe("/login");
		expect(wrapper.find(Link).text()).toEqual("Log in");
	});

	test("renders link to home when logged in", () => {
		store.dispatch(setUser({ username: "testName" }));
		const wrapper = mount(wrappedView);
		expect(wrapper.find(Link).prop("to")).toBe("/");
		expect(wrapper.find(Link).text()).toEqual("testName");
	});

	test("renders title to document.title", () => {
		const wrapper = mount(wrappedView);
		expect(document.title).toEqual(`DnD App | Test`);
	});
});
