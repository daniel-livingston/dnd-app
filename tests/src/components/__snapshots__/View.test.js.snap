// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<View> renders correctly 1`] = `
<div
  className="content"
>
  <div
    className="content__header"
  >
    <div
      className="content__header__info"
    >
      <h1
        className="content__header__title"
      >
        Test title
      </h1>
      <p
        className="content__header__description"
      >
        Test description
      </p>
    </div>
    <Link
      className="link"
      to="/login"
    >
      Log in
    </Link>
  </div>
  <div
    className="content__body"
  />
</div>
`;
